tower/src/util/try_with.rs

```diff
@@ -0,0 +1,70 @@
+use futures_util::future::{ready, Either, Ready};
+use std::task::{Context, Poll};
+use tower_layer::Layer;
+use tower_service::Service;
+
+/// Service returned by the [`try_with`] combinator.
+///
+/// [`try_with`]: crate::util::ServiceExt::try_with
+#[derive(Debug)]
+pub struct TryWith<S, F> {
+    inner: S,
+    f: F,
+}
+
+impl<S, F> TryWith<S, F> {
+    /// Creates a new [`TryWith`] service.
+    pub fn new(inner: S, f: F) -> Self {
+        TryWith { inner, f }
+    }
+}
+
+impl<S, F, R1, R2> Service<R1> for TryWith<S, F>
+where
+    S: Service<R2>,
+    F: FnOnce(R1) -> Result<R2, S::Error> + Clone,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = Either<S::Future, Ready<Result<S::Response, S::Error>>>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), S::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, request: R1) -> Self::Future {
+        match (self.f.clone())(request) {
+            Ok(ok) => Either::Left(self.inner.call(ok)),
+            Err(err) => Either::Right(ready(Err(err))),
+        }
+    }
+}
+
+/// A [`Layer`] that produces a [`TryWith`] service.
```

Comment:
Does `TryWith`  need a ref?

Reply:
["Docs.rs automatically builds crates' documentation released on crates.io using the nightly release of the Rust compiler".](https://docs.rs/about)

I've been testing it under `cargo +nightly doc --all-features --open` and everything seems to behave as intended.

Reply:
Okay, if these render correctly on docs.rs, that seems right to me --- I didn't realize that. Disregard the rest of my review! :)

Reply:
They seem to work in my experience. The one "gotcha" that's caught me out is that when you're working in a workspace locally and referencing those crates it works fine on `cargo +nightly doc`, but when you upload them individually to crates.io you'll lose your nice links via https://github.com/rust-lang/rust/issues/65983